### PR TITLE
[ML] Simplify the Inference Ingest Processor configuration

### PR DESCRIPTION
**Pull request by davidkyle**
Introduces a new way to configure which fields will be passed for inference and where to write the the results to.
- Removes the need to use a `field_map` to rename a field
- The output location is configured in 1 place next to the input
- Multiple fields from the ingest document can be selected for inference

The new configuration does not support [Data frame analytics](https://www.elastic.co/guide/en/kibana/current/xpack-ml-dfanalytics.html), those models accept the full ingest document not specific fields.  

### Example: Process the field `body` and write the results to `body_tokens`
```
{
    "processors": [
        {
            "inference": {
                "model_id": "elser_model",
                "input_output": [
                    {
                        "input_field": "body",
                        "output_field": "body_tokens"
                    }
                ]
            }
        }
    ]
}
```

### Example: Select multiple input fields can be specified as:
```
{
    "processors": [
        {
            "inference": {
                "model_id": "elser_model",
                "input_output": [
                    {
                        "input_field": "body",
                        "output_field": "body_tokens"
                    },
                    {
                        "input_field": "title",
                        "output_field": "title_tokens"
                    },
                ]
            }
        }
    ]
}
```

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/inference/InferenceResults.java [hunk 1]
```diff
@@ -24,16 +25,58 @@ static void writeResult(InferenceResults results, IngestDocument ingestDocument,
...
+     * @return Map representation of the InferenceResult
+     */
     Map<String, Object> asMap();
 
+    /**
+     * Convert to a map placing the inference result in {@code outputField}
+     * @param outputField Write the inference result to this field
+     * @return Map representation of the InferenceResult
+     */
+    Map<String, Object> asMap(String outputField);
```
[the remark this answers is not in the document]
Comment (jonathan-buttner) [reply server/src/main/java/org/elasticsearch/inference/InferenceResults.java:79]: Maybe this isn't feasible but the code here:

```
InferenceResults.writeResultToField(
                    response.getInferenceResults().get(i),
                    ingestDocument,
                    inputs.get(i).outputBasePath(),
                    inputs.get(i).outputField,
                    response.getId() != null ? response.getId() : modelId,
                    i == 0
```

So I think what I'm suggesting whoever is creating the inference results class that is eventually being placed in the response and then access here with `response.getInferenceResults().get(i)` would need to pass in the corresponding output field.

**Review by jonathan-buttner — APPROVED** [review]
Left a couple comments, I'm not sure they're worth keeping the PR from being merged though.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/results/ErrorInferenceResults.java [hunk 2]
```diff
@@ -74,6 +74,12 @@ public Map<String, Object> asMap() {
         return asMap;
     }
 
+    @Override
+    public Map<String, Object> asMap(String outputField) {
+        // errors do not have a result
```
[the remark this answers is not in the document]
Comment (davidkyle) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/results/ErrorInferenceResults.java:79]: `RawInferenceResults` is used internal by the tree ensembles, it should never be seen by the outside world which is why it throws. The comment here isn't very clear but the idea is that we don't write the error message to the result field (which might be mapped to a dense vector when written to the index), instead it goes in a different field

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessor.java [hunk 3]
```diff
@@ -82,39 +85,84 @@ public class InferenceProcessor extends AbstractProcessor {
...
         InferenceAuditor auditor,
         String tag,
         String description,
         String targetField,
         String modelId,
         InferenceConfigUpdate inferenceConfig,
-        Map<String, String> fieldMap
+        Map<String, String> fieldMap,
+        List<Factory.InputConfig> inputs,
+        boolean configuredWithInputsFields
```
[the remark this answers is not in the document]
Comment (davidkyle) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessor.java:148]: ++ good idea

**Pull request merged**